### PR TITLE
Adds rattler-build recipe compatibility (and adds jolt-physics)

### PR DESCRIPTION
**Pull request by baszalmstra**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

This PR adds support for the new recipe format and adds the first such recipe for https://github.com/jrouwe/JoltPhysics

We already did a previous attempt with https://github.com/conda-forge/staged-recipes/pull/27008 which we reverted because of some issues with conda-smithy. A new release is available which should fix this issue.

### CODE REVIEW

**Comment by baszalmstra** [comment]
Im unsure what is wrong with the WIndows build? There doesnt seem to be an actual error?

File: .ci_support/build_all.py
Comment (h-vetinari) [inline .ci_support/build_all.py:1]: Please remove all the formatting changes from this file, as it makes the actual changes unreviewable. Or at least, separate out the formatting into a separate commit before or after.

**Review by h-vetinari — COMMENTED** [review]
This PR does too many things IMO.
* format `build_all.py`
* add rattler-build support
* add a new rattler example recipe
* add a genuinely new recipe

Proving the functionality would already with actually building the example recipe (perhaps with some additional safeguard that we don't create a feedstock), and then a follow-up can add jolt.

W.r.t. to the error on windows, it seems to be failing while trying to call `inspect_artifacts`.
```
Inspecting artifacts
"inspect_artifacts needs conda-forge-ci-setup >=4.6.0"
##[error]Cmd.exe exited with code '1'.
```

**Comment by baszalmstra** [comment]
Thanks for checking out this PR @h-vetinari 

Note that this was already reviewed in https://github.com/conda-forge/staged-recipes/pull/27008 by both @jaimegp and @wolfv  but happy to make more changes.

The problem with the current setup is that there is no easy way to test this. The creation of the feedstock is performed by another repository so without merging this that code won't run and that code won't actually create a feedstock if there is no recipe, therefor we must add a recipe. 

The `build-all.py` script is tested currently by running the CI, if that turns green it should be good to go. I don't see why this recipe should be different from testing the create feedstock code.

The example recipe is there as an example similar to the existing example recipe. I would be happy to add it in another PR but in the previous round of reviews, it was mentioned that this was missing from the PR.

I undid the formatting.

File: recipes/example-new-recipe/recipe.yaml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  #   - win
+  script: python -m pip install . -vv
+  number: 0
+
+requirements:
+  build:
+    # If your project compiles code (such as a C extension) then add the required compilers as separate entries here.
+    # Compiler names include 'c', 'cxx' and 'fortran', among others.
+    - ${{ compiler('c') }}
+    - ${{ stdlib('c') }}  # If you need a compiler, add the `stdlib` equivalent too
```
Comment (h-vetinari) [inline recipes/example-new-recipe/recipe.yaml:56]: This makes it sound like other compilers (fortran, rust, cxx) also need an equivalent stdlib, but this is not the case. The C stdlib is unfortunately special in many ways (we've iterated on this with Wolf a couple of times, I don't have a better solution for now). I'd write
```suggestion
    - ${{ stdlib('c') }}  # If you need any compiler, add the C standard library ("stdlib") too
```

File: recipes/example-new-recipe/recipe.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  # If getting the source from GitHub, remove the line above,
+  # uncomment the line below, and modify as needed. Use releases if available:
+  # url: https://github.com/simplejson/simplejson/releases/download/${{ version }}/simplejson-${{ version }}.tar.gz
+  # and otherwise fall back to archive:
+  # url: https://github.com/simplejson/simplejson/archive/v${{ version }}.tar.gz
+  sha256: d58439c548433adcda98e695be53e526ba940a4b9c44fb9a05d92cd495cdd47f
+  # sha256 is the preferred checksum -- you can get it for a file with:
+  #  `openssl sha256 <file name>`.
+  # You may need the openssl package, available on conda-forge:
+  #  `conda install openssl -c conda-forge``
```
Comment (h-vetinari) [inline recipes/example-new-recipe/recipe.yaml:33]: I think it'd be helpful to suggest a variant that doesn't persist the file
```suggestion
  # sha256 is the preferred checksum -- you can get it for a file with:
  #  `openssl sha256 <file name>`,
  # or, in one go without keeping the file with:
  # `curl -L <url> | sha256
  # You may need the packages or openssl and/or curl, available on conda-forge:
  #  `conda install openssl curl -c conda-forge``
```
However, it should be noted that we're not currently [shipping](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fopenssl-3.3.1-h4bc722e_2.conda) any binaries for `sha256` (or `sha256sum`) in `openssl`, and so whether those get picked up depends entirely what else is on your path. Do you want to open an issue on the openssl feedstock to ship those?

File: .github/workflows/correct_directory.yml [hunk 3]
```diff
@@ -31,6 +31,7 @@ jobs:
                     '\nUnfortunately, the recipe was added directly in the `recipes` folder without its own subfolder.\n' +
                     'Please move the recipe file into a folder with the name of the package you want to submit.\n\n' +
                     'For example: if your recipe is currently under `recipes/<your_package>.yaml`, ' +
-                    'it should be moved to `recipes/<your_package>/meta.yaml`.\n' +
+                    'it should be moved to `recipes/<your_package>/meta.yaml` ' +
+                    'or, if you are using the new yaml based recipe format, `recipes/<your_package>/recipe.yaml`\n' +
```
Comment (h-vetinari) [inline .github/workflows/correct_directory.yml:35]: I think we need a better label than "new yaml based recipe format" (this partially applies to `example-new-recipe` as well), especially since the old one is already called `meta.yaml`. I know it's not "pure" yaml, but to a casual user this is just confusing. Why not something like "v2 of the recipe specification" or something along those lines?

File: .ci_support/build_all.py [hunk 4]
```diff
@@ -207,6 +242,50 @@ def build_folders(recipes_dir, folders, arch, channel_urls):
...
+    recipes_dir: str, platform, arch, channel_urls: list[str]
+):
+    config = get_config(arch, channel_urls)
+
+    # Remove the example recipes to ensure that they are not also build.
+    for example_recipe in EXAMPLE_RECIPE_FOLDERS:
+        rmtree(os.path.join(recipes_dir, example_recipe), ignore_errors=True)
+
+    # Determine the locations for the variant config files.
+    specs = OrderedDict()
```
Comment (h-vetinari) [inline .ci_support/build_all.py:255]: This is less a question for this PR and more a topic overall across our infrastructure, but at least I thought I'd note it here: https://github.com/conda-forge/conda-smithy/issues/2033

**Review by h-vetinari — COMMENTED** [review]
Thanks for removing the formatting changes! Looks pretty good at first glance already. I left some comments/suggestions.

**Comment by h-vetinari** [comment]
> Note that this was already reviewed in #27008 by both @jaimegp and @wolfv but happy to make more changes.

Yeah, fair enough, I don't want to complicate this process for you any further. Jaime would actually be a great person to ask for figuring out what (if anything) is going wrong with the artefact inspection. ;-)

File: .github/workflows/correct_directory.yml (hunk 3 above)
Comment (baszalmstra) [reply .github/workflows/correct_directory.yml:35]: There has been a whiole discussion going on an we setteld on recipe v1. Ill change this.

```suggestion
                    'or, if you are using the recipe v1 format, `recipes/<your_package>/recipe.yaml`\n' +
```

**Comment by baszalmstra** [comment]
The issue with `inspect-artifacts` should be fixed with https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/336

File: .ci_support/build_all.py (hunk 4 above)
Comment (baszalmstra) [reply .ci_support/build_all.py:255]:
```suggestion
    specs = {}
```

File: recipes/jolt-physics/recipe.yaml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            -S $SRC_DIR/Build
+        cmake --build . --target install
+
+about:
+  homepage: https://github.com/jrouwe/JoltPhysics
+  license: MIT
+  license_file: LICENSE
+  summary: A multi core friendly rigid body physics and collision detection library.
+  description: A multi core friendly rigid body physics and collision detection library. Written in C++. Suitable for games and VR applications. Used by Horizon Forbidden West. 
+  documentation: https://xtensor.readthedocs.iohttps://jrouwe.github.io/JoltPhysics/
```
Comment (jaimergp) [inline recipes/jolt-physics/recipe.yaml:58]:
```suggestion
  documentation: https://jrouwe.github.io/JoltPhysics/
```

File: recipes/example-new-recipe/recipe.yaml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    - ${{ compiler('c') }}
+    - ${{ stdlib('c') }}  # If you need any compiler, add the C standard library ("stdlib") too
+  host:
+    - python
+    - pip
+  run:
+    - python
+
+tests:
+  # Some packages might need a `test/commands` key to check CLI.
```
[the remark this answers is not in the document]
Comment (wolfv) [reply recipes/example-new-recipe/recipe.yaml:66]: True, that comment looks wrong. It is also `script` not command in the test section
